collector/src/main/java/com/bitmovin/analytics/retryBackend/RetryBackend.kt

```diff
@@ -0,0 +1,162 @@
+package com.bitmovin.analytics.retryBackend
+
+import android.os.Handler
+import android.os.SystemClock
+import android.util.Log
+import com.bitmovin.analytics.data.AdEventData
+import com.bitmovin.analytics.data.Backend
+import com.bitmovin.analytics.data.EventData
+import com.bitmovin.analytics.utils.Util.MAX_RETRY_SAMPLES
+import com.bitmovin.analytics.utils.Util.MAX_RETRY_TIME
+import java.io.IOException
+import java.lang.Exception
+import java.net.ConnectException
+import java.net.SocketTimeoutException
+import java.net.UnknownHostException
+import java.util.Calendar
+import java.util.Date
+import java.util.concurrent.locks.ReentrantLock
+import kotlin.math.pow
+import okhttp3.Call
+import okhttp3.Callback
+import okhttp3.Response
+import okhttp3.internal.http2.StreamResetException
+
+class RetryBackend(private val next: Backend, val handler: Handler) : Backend {
+
+    private val TAG = "RetryBackend"
+    private val lock = ReentrantLock()
+    private var retrySamplesSet = sortedSetOf<RetrySample>()
+    private var retryToken: Date? = null
+
+    override fun send(eventData: EventData, callback: Callback?) {
+        scheduleSample(RetrySample(eventData, null, 0, Date(), 0))
+    }
+
+    override fun sendAd(eventData: AdEventData, callback: Callback?) {
+        scheduleSample(RetrySample(null, eventData, 0, Date(), 0))
+    }
+
+    private fun scheduleSample(retrySample: RetrySample) {
+
+        Log.d(TAG, "sending sample retry ${retrySample.retry}")
+
+        if (retrySample.eventData != null) {
+            retrySample.eventData.retry = retrySample.retry
+            this.next.send(retrySample.eventData, object : Callback {
+
+                override fun onFailure(call: Call, e: IOException) {
+                    if (e is SocketTimeoutException || e is ConnectException || e is StreamResetException || e is UnknownHostException) {
+                        addSample(retrySample)
+                        call.cancel()
+                    }
+                }
+
+                override fun onResponse(call: Call, response: Response) {
+                }
+            })
+        } else if (retrySample.adEventData != null) {
+            retrySample.adEventData.retry = retrySample.retry
+            this.next.sendAd(retrySample.adEventData, object : Callback {
+                override fun onFailure(call: Call, e: IOException) {
+
+                    if (e is SocketTimeoutException) {
+                        addSample(retrySample)
+                    }
+                }
+
+                override fun onResponse(call: Call, response: Response) {
+                    processQueuedSamples()
+                }
+            })
+        }
+    }
+
+    @Synchronized
+    fun processQueuedSamples() {
+        try {
+            if (retrySamplesSet.size > 0) {
+                val nextScheduledTime = retrySamplesSet.first().scheduledTime
+
+                if (retryToken != nextScheduledTime) {
```

Comment:
what happens if two samples are scheduled at the same time?

Reply:
That should be ok, because then this condition will not be executed. Then once first sample gets sent again `token` will be set to `null`. Then this `processQueuedSamples` method will be called again and `nextScheduleTime` will be from other sample that had same time, and that sample will be scheduled to be immediately sent.